### PR TITLE
Updates to balred() and gram()

### PR DESCRIPTION
**Pull request by mdclemen**
balred() now supports "matchdc" and will handle unstable systems and will accept a list of ORDERS, which will return a list of reduced order systems. gram() now does Cholesky factored gramians if desired. Requires my updates to [slycot](https://github.com/python-control/Slycot/pull/2).

### CODE REVIEW

File: control/statefbk.py [hunk 1]
```diff
@@ -347,6 +347,9 @@ def gram(sys,type):
     #Check for ss system object
     if not isinstance(sys,statesp.StateSpace):
         raise ValueError("System must be StateSpace!")
+    gramtypes = ['c', 'o', 'cf', 'of']
```
Comment (slivingston) [inline control/statefbk.py]: Why do you declare `gramtypes` as a variable? It only appears to be used in the conditional expression on the next line, so the motivation for allocating a variable with function-level scope is not apparent to me.

File: control/modelsimp.py [hunk 2]
```diff
@@ -50,7 +50,7 @@
 from .statesp import StateSpace
 from .statefbk import gram
 
-__all__ = ['hsvd', 'balred', 'modred', 'era', 'markov', 'minreal']
+__all__ = ['hsvd', 'balred', 'modred', 'era', 'markov', 'minreal',]
```
Comment (slivingston) [inline control/modelsimp.py]: Why did you add a trailing `,` in this list expression? It is not a syntax error, but it is not consistent with the current style of the `control` package.

Comment (mdclemen) [inline control/modelsimp.py]: I forgot to delete the trailing ','.

File: control/statefbk.py (hunk 1 above)
Comment (mdclemen) [inline control/statefbk.py]: I don't really remember, I wrote that function a few months ago. If you think there is a better way to do things, then by all means change it.

Comment (slivingston) [inline control/statefbk.py]: For the reason that I gave above, moving the list into the conditional expression improves code quality, and thus, I recommend the change. However, it is not critical for accepting this pull request.

Just to be clear, I am belatedly providing review of your proposed changes now.

File: control/modelsimp.py (hunk 2 above)
Comment (slivingston) [inline control/modelsimp.py]: Can you delete it, then? I am writing additional comments now, so you might want to wait before rebasing until I send them all.

**Comment by slivingston** [comment]
In the first commit (<https://github.com/python-control/python-control/pull/118/commits/cc4b5c75e676b86dd3ba16850c6a7361b0f30f0b>) of this pull request, the message and author are incorrectly copied from cdd3e73838cea0b26a7155d0b2084aa698e5eefd. When I skim the diff, it appears that you applied changes as part of your own work, but then, those were combined with the patch corresponding to commit cdd3e73838cea0b26a7155d0b2084aa698e5eefd. Can you clean it up, in particular remove the re-applied pieces of commit cdd3e73838cea0b26a7155d0b2084aa698e5eefd?

**Comment by slivingston** [comment]
It appears that substantial amounts of code are added in the first commit and then deleted or modified in the second. Can you add details in the commit messages to explain this transition? Alternatively, can you squash changes into a single commit?

Those are all of my questions and requests for now. I will wait until you respond or rebase before continuing to review.

**Comment by mdclemen** [comment]
I made changes to the code that you recommended, but my git skills might not be good enough for the other things you want me to do.

**Comment by slivingston** [comment]
@mdclemen Can I have your permission to create a commit that has your changes from this PR and lists you as author (and me as committer)? This way I can make the rebasing adjustments that I requested above. If I do this, how do you prefer your name to be written? E.g., I can use "clementm" as in your commit <https://github.com/python-control/python-control/pull/118/commits/f74e66552aa30645166f62fa4e5c115a196f85ca>, or your name as it appears in your publications.

Alternatively, we can arrange to meet in IRC to work through the `rebase` process together.

File: control/modelsimp.py [hunk 3]
```diff
@@ -211,26 +218,46 @@ def balred(sys, orders, method='truncate'):
         of systems)
     method: string
         Method of removing states, either ``'truncate'`` or ``'matchdc'``.
+    alpha: float
+        Specifies the alpha-stability boundary for the eigenvalues
```
Comment (slivingston) [inline control/modelsimp.py]: This text description of `alpha` appears to be copied from AB09MD.f of the SLICOT source code. Is that true? SLICOT is under the GNU GPL, whereas we want to keep this `control` package under [the BSD license](https://github.com/python-control/python-control/blob/c1f5a9e06fa1cce06a859edcfb41b257ab772372/LICENSE).

Comment (slivingston) [inline control/modelsimp.py]: If the text was indeed copied, can you create a new description of the `alpha` parameter? An easy solution for now is to write a brief summary and indicate where more information can be found (the documentation of SLICOT).

Comment (slivingston) [inline control/modelsimp.py]: The original text from SLICOT to which I am referring is in [lines 89-96 of AB09MD.f](https://github.com/python-control/Slycot/blob/b01430c0f5117e1d6b6fd1717e04fe765a37ebf2/slycot/src/AB09MD.f#L89-L96)

File: control/modelsimp.py [hunk 4]
```diff
@@ -197,10 +197,17 @@ def modred(sys, ELIM, method='matchdc'):
...
     return rsys
 
-def balred(sys, orders, method='truncate'):
+def balred(sys, orders, method='truncate', alpha=None):
     """
     Balanced reduced order model of sys of a given order.
     States are eliminated based on Hankel singular value.
+    If sys has unstable modes, they are removed, the
+    balanced realization is done on the stable part, then
+    reinserted IAW reference below.
```
Comment (slivingston) [inline control/modelsimp.py]: What is "IAW"?

**Comment by mdclemen** [comment]
In Accordance With

File: control/modelsimp.py (hunk 4 above)
Comment (slivingston) [inline control/modelsimp.py]: can you write "in accordance with" explicitly? The abbreviation might not be widely recognized.

I can also make the change after merging.

**Comment by slivingston** [comment]
I resolved the merge conflicts, but before I apply them here, can you first address my question about the description of the `alpha` parameter?

File: control/modelsimp.py (hunk 3 above)
Comment (mdclemen) [inline control/modelsimp.py]: I did in fact copy the description of `alpha` from AB09MD.f.

**Comment by slivingston** [comment]
Should we keep the original call to `AB09AD` when the system is stable? Otherwise (if `method='truncate'` and the system is not stable), your newly introduced call to `AB09MD` is used.

The Fortran implementations of these two subroutines are different, so I am motivated to understand them more before removing the one dedicated to the case of stable systems.

**Comment by slivingston** [comment]
Once we decide actions regarding `AB09AD` and the description of the `alpha` parameter, then this PR will be ready for merging. Thanks for your good work on the `gram` and `balred` functions!

**Comment by mdclemen** [comment]
I don't know why the SLICOT authors decided to include a user defined boundary for what constitutes unstable eigenvalues but essentially `alpha` allows the user to redefine the boundary for a stable eigenvalue.

**Comment by mdclemen** [comment]
@slivingston let me know what you think of my updates.

File: control/modelsimp.py [hunk 5]
```diff
@@ -219,12 +219,12 @@ def balred(sys, orders, method='truncate', alpha=None):
...
-        of the state dynamics matrix A. For a continuous-time
-        system, alpha <= 0 is the boundary value for the real parts
-        of eigenvalues, while for a discrete-time system, 0 <= alpha <= 1
-        represents the boundary value for the moduli of eigenvalues.
-        The alpha-stability domain does not include the boundary.
+        Redefines the stability boundary for eigenvalues of the system matrix A.
+        By default for continuous-time systems, alpha <= 0 defines the stability
+        boundary for the real part of A's eigenvalues and for discrete-time
+        systems, 0 <= alpha <= 1 defines the stability boundary for the modulus
+        of A's eigenvalues.See SLICOT routines AB09MD and AB09ND for more
```
Comment (slivingston) [inline control/modelsimp.py]: While resolving the merge conflict, I added a space here after the `.`. So, now it reads "...eigenvalues. See SLICOT...".

**Comment by slivingston** [comment]
Your updates are good. Thanks for addressing my comments.
